feat: implement jupiter as a swapper

NeOMakinG opened the pull request:
## Description
Implementing the Jupiter swapper:

- We are delegating the fee estimate at quote time to jupiter as they are building the TX instructions, it's safer and more opti in term of execution time
- We need to use the wrapped SOL assetId instead of the SOL, jupited is unwrapping and wrapping directly in their program
- Added affiliate fees account and BPS
- Selecting wrapped sol shouldn't be possible as Jupiter is using it as underlying asset, if we send the wrapped sol balance to the Jupiter API, it will consider we want to swap SOL (as a follow up, we could display somewhere on the wrapped sol asset page or swapper that the user can unwrap these wrapped sol by clicking on a button)
- ⚠️  Waiting for the DAO fee address for affiliate

<!-- Please describe your changes -->

## Issue (if applicable)

<!-----------------------------------------------------------------------------
If applicable, please link to the github issue and put `closes #XXXX` in your comment to auto-close the issue that your PR fixes.
------------------------------------------------------------------------------>

closes #8017

## Risk
High because swapper but behind a feature flag
> High Risk PRs Require 2 approvals

<!-----------------------------------------------------------------------------
Outline the scope of your changes and the risk associated with them. You must use your discretion as an engineer to determine the potential impact of your changes.

WARNING: If your PR introduces a new on-chain transaction or modifies an existing one, please add the "High-Risk" label to this PR and ask for 2 reviewers to approve it before merging.

E.g. an upgrade to `hdwallet` or core state management would be considered higher risk, and might require a full regression test. UI or isolated view changes, or something behind a feature flag may have near zero risk. Small bug fixes might require testing isolated to the specific fix.
------------------------------------------------------------------------------>

> What protocols, transaction types, wallets or contract interactions might be affected by this PR?

## Testing
- Try to do a swap from SOL to any SPL token
- Try to do a swap from SPL to SPL
- Try to do a swap from SPL to SOL
- Try all of those but with a manual receive address (SPL to SOL shouldn't provide a quote because it's not effectively supported by Jupiter (it leave some wSOL in the receive address which is not acceptable for us))
- Please try to swap to fresh accounts, and send again so it already have an associated token account, it's 2 different mechanisms, also try to swap a token you never had in your wallet
- Try to swap some Token2022 tokens too (catwifhat is one of those), specially with a manual receive address
- Smoke test swapper but shouldn't change anything to the rest

<!-----------------------------------------------------------------------------
We treat every PR to be merged with the same scrutiny as if we were merging directly to production.

Your PR will not be merged if you do not complete the sections below for our engineering and operations teams to test.
------------------------------------------------------------------------------>

### Engineering
n/a
<!-----------------------------------------------------------------------------
Include sufficient information here for an engineer to test your PR. This may include how to test locally, in a built environment, changes to infrastructure etc.
------------------------------------------------------------------------------>

### Operations
n/a
- [x] :checkered_flag: My feature is behind a flag and doesn't require operations testing (yet)

<!-----------------------------------------------------------------------------
If your changes have a user-facing impact, describe how a non-technical QA team can functionally test your changes in a preview environment.

If they are not user-facing please describe how to test for any regressions that may occur.
------------------------------------------------------------------------------>

## Screenshots (if applicable)
## SOL to SPL
<img width="474" alt="image" src="https://github.com/user-attachments/assets/3fcdfb98-7147-4e00-8b5e-ed4e419262a7">

## SPL to SOL
<img width="473" alt="image" src="https://github.com/user-attachments/assets/c540e0b6-674b-4e22-88ae-e2cb973c8b26">
https://jam.dev/c/ce54f3bb-43f1-4e1f-9f44-c8b44a19fb73

## SPL to SPL
⚠️ It was a high volatility SPL token, I had to increase the slippage tolerange
https://explorer.solana.com/tx/3jfq6B3z44Y7ffJxfVm2wRAKFJMba8fANUitHSguzmiKgb8NYfyRfWRHyqvokxHbRe8Qi5mUyuESoAuUEZzwTau5
<img width="473" alt="image" src="https://github.com/user-attachments/assets/9b74cee9-4ad3-4f9e-98c8-d5440402afba">
https://jam.dev/c/16c8e5f0-d5e2-4530-865b-b324a6980e25

## TX history
<img width="794" alt="image" src="https://github.com/user-attachments/assets/2873c7c3-d742-4756-9665-a54e2f23df69">
<img width="824" alt="image" src="https://github.com/user-attachments/assets/b1158875-9f01-4b0c-9ae5-7e74109119a2">
<img width="818" alt="image" src="https://github.com/user-attachments/assets/d894c4d0-17b2-48a5-b3ac-f33cf0195b8b">


## Txs links
https://explorer.solana.com/tx/23xAVVo18sgD2ykyqH5Kx9C37j6bLNjbFXi2rRW4ge5YArm1ywsTFGXKzSRmAQV6Dx13GnXZP5W7f1UNjmrAqDau
https://explorer.solana.com/tx/5k2r33Dd2oPKf5NESmQZz2G2fQ23wSy7xfL82PhCKwZuT5kojmynuHrVDy3uzhsf2UFq13wKjcvJMr61cdsGDGBL
https://explorer.solana.com/tx/2ucg29ZHH58uVZ9MCbE3VCCfwgHVdcozcxG4NfELDmTQTew4T6HQKh6E8eBx7Xr12K2xjZEEHpHDfdi1fQxg7Tyv
https://explorer.solana.com/tx/4XgfBYeqrn5pq2vTGmYrktYmgii1EiUXR7vZNzmULNNQgePMQVMKZsVWPbChURtUrYN9w8cD2aFteVtP1fCet8LV

NeOMakinG commented on packages/swapper/src/swappers/JupiterSwapper/swapperApi/getTradeQuote.ts:
```diff
@@ -0,0 +1,221 @@
+
+  const maybeQuoteResponse = await getJupiterQuote({
+    apiUrl: jupiterUrl,
+    sourceAsset: sellAsset.assetId === solAssetId ? wrappedSolAssetId : sellAsset.assetId,
+    destinationAsset: buyAsset.assetId === solAssetId ? wrappedSolAssetId : buyAsset.assetId,
```
Jupiter use wrapped SOL as an asset to swap with SOL, it auto unwrap/wrap in the instructions, this is how it works!

NeOMakinG replied:
Very nice catch, It's supported but I need to disable the auto wrap feature in that case, updated in [4e76e81](https://github.com/shapeshift/web/pull/8120/commits/4e76e8115223d91bd602f7d9ceed80c94c77ea15) , if the buyAsset is wrapped sol, don't unwrap automatically

NeOMakinG replied:
But there are magic somewhere that prevent from selecting wapped sol for some reasons

gomesalexandre replied:
LGTM! Will take a look at selection when testing this at runtime

NeOMakinG commented on packages/swapper/src/swappers/JupiterSwapper/endpoints.ts:
```diff
@@ -0,0 +1,144 @@
+
+    const adapter = assertGetSolanaChainAdapter(step.sellAsset.chainId)
+
+    const isManualReceiveAddress = tradeQuote.receiveAddress !== from
```
Manual address is very specific, it has multiple edge cases that I had to handle:

- We are not using wallet address but token ATA, so we need to create these ATA if they don't exist, and use them in the jupiter swap instructions endpoints
- If it's Token2022 or SPL, this ATA derivation changes
- If we send SOL with manual receive address, we don't support it because Jupiter only supports sending wSOL under the hood with external address

gomesalexandre replied:
Renamed to `isCrossAccountTrade` in previous commits and plumbing cleaned up, closing this

NeOMakinG commented on packages/swapper/src/swappers/JupiterSwapper/endpoints.ts:
```diff
@@ -0,0 +1,144 @@
+      toAddress: isManualReceiveAddress ? destinationTokenAccount?.toString() : undefined,
+      rawQuote: tradeQuote.rawQuote,
+      // Shared account is not supported for simple AMMs
+      useSharedAccounts:
```
Another edge case:

Shared accounts should be used for manual address if there are multiple routes, because jupiter doesn't support shared accounts for single hop simple AMM

And we need to use shared accounts in the context of sending to manual receive address so funds are not stuck in a subaccount of the sending address

NeOMakinG commented on packages/swapper/src/swappers/JupiterSwapper/endpoints.ts:
```diff
@@ -0,0 +1,144 @@
+    const maybeSwapResponse = await getJupiterSwapInstructions({
+      apiUrl: jupiterUrl,
+      fromAddress: from,
+      toAddress: isManualReceiveAddress ? destinationTokenAccount?.toString() : undefined,
```
Straight but if manualreceive address, we add token ATA to this instead of account address

NeOMakinG commented on packages/swapper/src/swappers/JupiterSwapper/endpoints.ts:
```diff
@@ -0,0 +1,144 @@
+            value: '0',
+            chainSpecific: {
+              from,
+              addressLookupTableAccounts: swapResponse.addressLookupTableAddresses,
```
I added this or we will hit the Solana TX max length, we absolutely need this to be able to send the whole TX

NeOMakinG commented on packages/swapper/src/swappers/JupiterSwapper/endpoints.ts:
```diff
@@ -0,0 +1,144 @@
+    )
+
+    const buildSwapTxInput: BuildSendApiTxInput<KnownChainIds.SolanaMainnet> = {
+      to: '',
+      from,
+      value: '0',
```
We absolutely never send any values as a regular send, but we always use jupiter instructions, so we zero out every values as we were sending an SPL token

NeOMakinG commented on packages/chain-adapters/src/solana/SolanaChainAdapter.ts:
```diff
@@ -433,40 +462,49 @@ export class ChainAdapter implements IChainAdapter<KnownChainIds.SolanaMainnet>
+    instruction?: TransactionInstruction
+    destinationTokenAccount: PublicKey
+  }> {
+    const accountInfo = await this.connection.getAccountInfo(new PublicKey(tokenId))
+
+    const TOKEN_PROGRAM =
+      accountInfo?.owner.toString() === TOKEN_2022_PROGRAM_ID.toString()
+        ? TOKEN_2022_PROGRAM_ID
+        : TOKEN_PROGRAM_ID
```
This is some of the meaty logic:

Sending Token2022 tokens were probably not working because of this, we were always using the SPL token programId (ex: sending catwifhat to someone already owning it or who never owned it)

Now we support both (I don't know any other protocols, I think it's the 2 used globally)

Needed for swapping to a Token2022 token using an external receive address (another edge case)

woodenfurniture commented on .env.base:
```diff
@@ -68,6 +68,7 @@ REACT_APP_FEATURE_LIFI_SWAP=true
 REACT_APP_FEATURE_THOR_SWAP=true
 REACT_APP_FEATURE_THOR_SWAP_STREAMING_SWAPS=true
 REACT_APP_FEATURE_ZRX_SWAP=true
+REACT_APP_FEATURE_JUPITER=false
```
```suggestion
REACT_APP_FEATURE_JUPITER_SWAP=false
```

NeOMakinG replied:
Done in [e4891d6](https://github.com/shapeshift/web/pull/8120/commits/e4891d6256da07fae14a9f7eaa0e89ba7aefbe0d)

woodenfurniture commented on src/state/slices/preferencesSlice/preferencesSlice.ts:
```diff
@@ -72,6 +72,7 @@ export type FeatureFlags = {
   ChainflipDca: boolean
   PublicTradeRoute: boolean
   ThorFreeFees: boolean
+  Jupiter: boolean
```
```suggestion
  JupiterSwap: boolean
```

NeOMakinG replied:
Done in [e4891d6](https://github.com/shapeshift/web/pull/8120/commits/e4891d6256da07fae14a9f7eaa0e89ba7aefbe0d)

woodenfurniture reviewed (changes requested):
First pass, looking overall pretty sane. Will wait to resolve these before going deep

NeOMakinG commented on packages/chain-adapters/src/utils/solanaUtils.ts:
```diff
@@ -0,0 +1,9 @@
+
+import { solanaChainIds } from '../solana/SolanaChainAdapter'
+
+export const isSolanaChainId = (
```
Sealevel is the underlying protocol of parallelizing smart contracts but SVM (Solana Virtual Machine) looks more near than what we call EVM (https://solana.com/developers/evm-to-svm/smart-contracts)

NeOMakinG replied:
I would call it SVM as they do, and as we do for EVM, changed in [e4891d6](https://github.com/shapeshift/web/pull/8120/commits/e4891d6256da07fae14a9f7eaa0e89ba7aefbe0d)

gomesalexandre replied:
See review in https://github.com/shapeshift/web/pull/8120#pullrequestreview-2456971330 - `isSvmChainId` *is* more correct indeed, however, it may or may not be an early abstraction.
Totally happy with keeping the SVM terminology/abstraction if we wish to, doesn't really hurt to, but we should just be cognicent that chances are that we will never add other SVM ChainIds, meaning that this is one additional layer to reason about here vs. just checking for explicit `SolanaChainId` across the board

NeOMakinG replied:
We kicked this logic from the PR

NeOMakinG commented on packages/swapper/src/swappers/JupiterSwapper/models/AccountMeta.ts:
```diff
@@ -0,0 +1,84 @@
+// @ts-nocheck
+/* tslint:disable */
+/* eslint-disable */
```
@woodenfurniture we already does this in multiple places, those types are generated OPEN API types, I imported only the bits we consume to avoid spamming our repository with unused bits as we did for Chainflip already

NeOMakinG commented on packages/swapper/src/swappers/JupiterSwapper/models/Instruction.ts:
```diff
@@ -0,0 +1,91 @@
+// @ts-nocheck
+/* tslint:disable */
+/* eslint-disable */
```
see https://github.com/shapeshift/web/pull/8120#discussion_r1853635914

NeOMakinG commented on packages/swapper/src/swappers/JupiterSwapper/models/AccountMeta.ts:
```diff
@@ -0,0 +1,84 @@
+// @ts-nocheck
+/* tslint:disable */
+/* eslint-disable */
+/**
+ * Jupiter API v6
+ * The core of [jup.ag](https://jup.ag). Easily get a quote and swap through Jupiter API.  ### Rate Limit We update our rate limit from time to time depending on the load of our servers. We recommend running your own instance of the API if you want to have high rate limit, here to learn how to run the [self-hosted API](https://station.jup.ag/docs/apis/self-hosted).  ### API Wrapper - Typescript [ ...
+ *
+ * The version of the OpenAPI document: 6.0.0
+ * 
+ *
+ * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
+ * https://openapi-generator.tech
+ * Do not edit the class manually.
+ */
```
Types are coming from https://github.com/jup-ag/jupiter-quote-api-node

Should we consume their npm module and remove the types from our repository instead? Initially I feel like it was better to import those types so we don't add an extra dependency but I'm open for discussion now

woodenfurniture replied:
Absolutely, if importing them is an option (it wasnt for cowswap due to cjs bundling shenanigans).

NeOMakinG commented on src/state/helpers.ts:
```diff
@@ -41,6 +51,8 @@ export const getEnabledSwappers = (
     [SwapperName.Chainflip]:
       Chainflip && (!isCrossAccountTrade || isCrossAccountTradeSupported(SwapperName.Chainflip)),
+    [SwapperName.Jupiter]:
+      Jupiter && (!isCrossAccountTrade || isCrossAccountTradeSupported(SwapperName.Jupiter)),
```
Done in [e4891d6](https://github.com/shapeshift/web/pull/8120/commits/e4891d6256da07fae14a9f7eaa0e89ba7aefbe0d)

NeOMakinG commented on react-app-rewired/headers/csps/jupiter.ts:
```diff
@@ -0,0 +1,5 @@
+import type { Csp } from '../types'
+
+export const csp: Csp = {
+  'connect-src': ['https://quote-api.jup.ag/'],
```
It makes total sense, CSP are not updated if we update the env variable

I would like to avoid fixing all the other cases, can we create an issue to fix all the other cases?

Jup handled in [e4891d6](https://github.com/shapeshift/web/pull/8120/commits/e4891d6256da07fae14a9f7eaa0e89ba7aefbe0d)

woodenfurniture replied:
Agree, lets not let this PR turn into "fix: shapeshift" 😅

NeOMakinG commented on packages/chain-adapters/src/solana/SolanaChainAdapter.ts:
```diff
@@ -57,6 +64,8 @@ import { toAddressNList, toRootDerivationPath } from '../utils'
 import { assertAddressNotSanctioned } from '../utils/validateAddress'
 import { microLamportsToLamports } from './utils'
 
+export const solanaChainIds = [KnownChainIds.SolanaMainnet] as const
```
I changed it in this PR but I feel like we are a bit off now

We might missed something while implementing the SolanaChainAdapter => All those logic bits are called `Solana`, while we could effectively name is `SVM`

We are not planning to support any more SVM but it exists and we might want to be already supporting this type of chain as you said

I won't change all the things we did in the past in this PR but happy to already adapt what I added so we start something

woodenfurniture replied:
Sounds good, and I may be completely off with this anyway. Appreciate the pragmatism here ser.

NeOMakinG commented on packages/swapper/src/swappers/JupiterSwapper/utils/helpers.ts:
```diff
@@ -0,0 +1,74 @@
+const JUPITER_TRANSACTION_MAX_ACCOUNTS = 54
+
+// const SHAPESHIFT_SOLANA_FEE_ACCOUNT = 'C7RTJbss7R1r7j8NUNYbasUXfbPJR99PMhqznvCiU43N'
+
```
Removed in [e4891d6](https://github.com/shapeshift/web/pull/8120/commits/e4891d6256da07fae14a9f7eaa0e89ba7aefbe0d) and add a todo for the fee account missing

NeOMakinG commented on packages/swapper/src/swappers/JupiterSwapper/utils/jupiterService.ts:
```diff
@@ -0,0 +1,20 @@
+import { SwapperName } from '../../../types'
+import { createCache, makeSwapperAxiosServiceMonadic } from '../../../utils'
+
+const maxAge = 15 * 1000 // 15 seconds ttl for all except cached to get fresh status updates
```
Updated in [e4891d6](https://github.com/shapeshift/web/pull/8120/commits/e4891d6256da07fae14a9f7eaa0e89ba7aefbe0d)

NeOMakinG commented on packages/swapper/src/swappers/JupiterSwapper/utils/constants.ts:
```diff
@@ -0,0 +1,11 @@
+
+import type { SupportedChainIds } from '../../../types'
+
+export const jupiterSupportedChainIds = [KnownChainIds.SolanaMainnet] as const
```
Changed in [e4891d6](https://github.com/shapeshift/web/pull/8120/commits/e4891d6256da07fae14a9f7eaa0e89ba7aefbe0d)

NeOMakinG commented on packages/swapper/src/swappers/JupiterSwapper/types.ts:
```diff
@@ -0,0 +1,3 @@
+import type { jupiterSupportedChainIds } from './utils/constants'
+
+export type JupiterSupportedChainId = (typeof jupiterSupportedChainIds)[number]
```
Removed in [e4891d6](https://github.com/shapeshift/web/pull/8120/commits/e4891d6256da07fae14a9f7eaa0e89ba7aefbe0d)

NeOMakinG commented on packages/types/src/base.ts:
```diff
@@ -48,6 +48,8 @@ export type UtxoChainId =
   | KnownChainIds.DogecoinMainnet
   | KnownChainIds.LitecoinMainnet
 
+export type SolanaChainId = KnownChainIds.SolanaMainnet
```
Updated in [e4891d6](https://github.com/shapeshift/web/pull/8120/commits/e4891d6256da07fae14a9f7eaa0e89ba7aefbe0d)

gomesalexandre replied:
@woodenfurniture see https://github.com/shapeshift/web/pull/8120#discussion_r1855992498, not sure what we think about that but I'd be inclined to say that this is an early abstraction and all we care about is *explicit* Solana ChainId, though don't think it hurts to keep it either for the sake of consistency so happy either way

woodenfurniture replied:
Yep, appreciate the pragmatism. Again, I may be completely off with the sealevel thing so lets not abstract early.

gomesalexandre commented on packages/swapper/src/swappers/JupiterSwapper/swapperApi/getTradeQuote.ts:
```diff
@@ -0,0 +1,231 @@
+
+      acc[feeAssetId] = {
+        requiresBalance: false,
+        amountCryptoBaseUnit: bnOrZero(route.swapInfo.feeAmount).toString(),
```
As a paranoia to not end up with a Frankenstein base unit amount that's actually a float
```suggestion
        amountCryptoBaseUnit: bnOrZero(route.swapInfo.feeAmount).toFixed(0),
```

NeOMakinG replied:
Changed in [4e76e81](https://github.com/shapeshift/web/pull/8120/commits/4e76e8115223d91bd602f7d9ceed80c94c77ea15)

gomesalexandre commented on packages/swapper/src/swappers/JupiterSwapper/swapperApi/getTradeQuote.ts:
```diff
@@ -0,0 +1,231 @@
+    receiveAddress,
+    potentialAffiliateBps: affiliateBps,
+    affiliateBps,
+    isStreaming: false,
```
nitpick: this is optional and not applicable to this domain so might as well remove this line

gomesalexandre commented on packages/swapper/src/swappers/JupiterSwapper/utils/helpers.ts:
```diff
@@ -0,0 +1,72 @@
+import { jupiterSupportedChainIds } from './constants'
+import { jupiterService } from './jupiterService'
+
+export const isSupportedChainId = (chainId: ChainId): chainId is ChainId => {
```
suggestion: narrow down return type here

NeOMakinG replied:
[4e76e81](https://github.com/shapeshift/web/pull/8120/commits/4e76e8115223d91bd602f7d9ceed80c94c77ea15)

gomesalexandre commented on packages/swapper/src/swappers/JupiterSwapper/utils/constants.ts:
```diff
@@ -0,0 +1,11 @@
+
+import type { SupportedChainIds } from '../../../types'
+
+export const jupiterSupportedChainIds = [KnownChainIds.SolanaMainnet] as ChainId[]
```
nitpick: 
```suggestion
export const jupiterSupportedChainIds: [KnownChainIds.SolanaMainnet] = [KnownChainIds.SolanaMainnet]
```

NeOMakinG replied:
Yes or we will have some issues with types deeper in the tree (try it out if you want)

gomesalexandre commented on packages/swapper/src/swappers/JupiterSwapper/utils/helpers.ts:
```diff
@@ -0,0 +1,72 @@
+  destinationAsset: string
+  commissionBps: string
+  amount: string
+  slippageBps?: string
```
nitpick: make this required as it will actually always be defined currently:

https://github.com/shapeshift/web/blob/f7317eeb8a0b22774a66a2bc84eb0f6a7dfcabeb/packages/swapper/src/swappers/JupiterSwapper/swapperApi/getTradeQuote.ts#L86-L89

But if it wasn't, we wouldn't catch it

NeOMakinG replied:
Made in [4e76e81](https://github.com/shapeshift/web/pull/8120/commits/4e76e8115223d91bd602f7d9ceed80c94c77ea15)

gomesalexandre commented on packages/swapper/src/swappers/JupiterSwapper/utils/helpers.ts:
```diff
@@ -0,0 +1,72 @@
+  rawQuote: unknown
+  toAddress?: string
+  useSharedAccounts: boolean
+  wrapAndUnwrapSol?: boolean
```
suggestion: remove me, we never pass it and always leverage `wrapAndUnwrapSol = true`, so can just pass:
`wrapAndUnwrapSol: true` directly in the `jupiterService.post()` call

NeOMakinG replied:
Not removed but consumed for the `wrappedSol` usecase in [4e76e81](https://github.com/shapeshift/web/pull/8120/commits/4e76e8115223d91bd602f7d9ceed80c94c77ea15)

gomesalexandre commented on packages/swapper/src/swappers/JupiterSwapper/utils/helpers.ts:
```diff
@@ -0,0 +1,72 @@
+  apiUrl: string
+  fromAddress: string
+  rawQuote: unknown
+  toAddress?: string
```
q: do we want to make this required and always pass it in (although maybe undefined) for paranoia?

gomesalexandre commented on packages/swapper/src/swappers/JupiterSwapper/utils/jupiterService.ts:
```diff
@@ -0,0 +1,20 @@
+import { SwapperName } from '../../../types'
+import { createCache, makeSwapperAxiosServiceMonadic } from '../../../utils'
+
+const maxAgeMillis = 15 * 1000 // 15 seconds ttl for all except cached to get fresh status updates
+const cachedUrls: string[] = []
```
This is copied over from Chainflip but doesn't look correct, no URLs are cached here

```suggestion
const maxAgeMillis = 15 * 1000
const cachedUrls: string[] = []
```

NeOMakinG replied:
Changed in [4e76e81](https://github.com/shapeshift/web/pull/8120/commits/4e76e8115223d91bd602f7d9ceed80c94c77ea15)

gomesalexandre commented on packages/swapper/src/swappers/utils/filterBuyAssetsBySellAssetId/filterBuyAssetsBySellAssetId.ts:
```diff
@@ -23,6 +23,19 @@ const _filterEvmBuyAssetsBySellAssetId = (
   })
 }
 
+const _filterSolanaBuyAssetsBySellAssetId = (
+  { assets, sellAsset }: BuyAssetBySellIdInput,
+  chainIdPredicate: ChainIdPredicate,
+): Asset[] => {
+  // evm only
+  if (!isSvmChainId(sellAsset.chainId)) return []
+
+  return assets.filter(buyAsset => {
+    // evm only AND chain id predicate with no arbitrum nova support for any swappers
+    return isSvmChainId(buyAsset.chainId) && chainIdPredicate(buyAsset.chainId, sellAsset.chainId)
+  })
+}
+
```
```suggestion
const _filterSolanaBuyAssetsBySellAssetId = (
  { assets, sellAsset }: BuyAssetBySellIdInput,
  chainIdPredicate: ChainIdPredicate,
): Asset[] => {
  // svm only
  if (!isSvmChainId(sellAsset.chainId)) return []

  return assets.filter(buyAsset => {
    // svm only AND chain id predicate
    return isSvmChainId(buyAsset.chainId) && chainIdPredicate(buyAsset.chainId, sellAsset.chainId)
  })
}
```

Though we should be able to remove this entirely after tackling the suggested changes in `filterBuyAssetsBySellAssetId`

NeOMakinG replied:
Totally removed in [4e76e81](https://github.com/shapeshift/web/pull/8120/commits/4e76e8115223d91bd602f7d9ceed80c94c77ea15)

gomesalexandre commented on packages/swapper/src/swappers/utils/filterBuyAssetsBySellAssetId/filterBuyAssetsBySellAssetId.ts:
```diff
@@ -36,3 +49,11 @@ export const filterCrossChainEvmBuyAssetsBySellAssetId = (
   return _filterEvmBuyAssetsBySellAssetId(input, crossChainIdPredicate)
 }
+
+export const filterSameChainSolanaBuyAssetsBySellAssetId = (
+  input: BuyAssetBySellIdInput,
+): Asset[] => {
+  const sameChainIdPredicate = (buyAssetChainId: ChainId, sellAssetChainId: ChainId): boolean =>
+    buyAssetChainId === sellAssetChainId
+  return _filterSolanaBuyAssetsBySellAssetId(input, sameChainIdPredicate)
+}
```
And this one should be removed too after tackling `filterBuyAssetsBySellAssetId` suggested changes

NeOMakinG replied:
Removed in [4e76e81](https://github.com/shapeshift/web/pull/8120/commits/4e76e8115223d91bd602f7d9ceed80c94c77ea15)

gomesalexandre commented on packages/swapper/src/types.ts:
```diff
@@ -422,13 +427,22 @@ export type CheckTradeStatusInput = {
 type TradeQuoteResult = Result<TradeQuote[], SwapErrorRight>
 export type TradeRateResult = Result<TradeRate[], SwapErrorRight>
 
+export type SolanaTransactionRequest = {
+  gasLimit: string
+  to: string
+  from: string
+  value: string
+  data: string
+}
+
```
Remove me, unused and not a Solana request

NeOMakinG replied:
Damn, removed in [4e76e81](https://github.com/shapeshift/web/pull/8120/commits/4e76e8115223d91bd602f7d9ceed80c94c77ea15)

gomesalexandre commented on packages/swapper/src/utils.ts:
```diff
@@ -318,3 +325,37 @@ export const isToken = (assetId: AssetId) => {
+}: {
+  txHash: string
+  accountId: AccountId | undefined
+  chainId: ChainId
```
suggestion: don't parametrize me. `chainId` makes sense for `checkEvmSwapStatus` since we're dealing with different EVM chains, but this method is called `checkSolanaSwapStatus`, not `checkSvmSwapStatus` so the `chainId` is guaranteed to always be solanaChainId

gomesalexandre commented on packages/types/src/base.ts:
```diff
@@ -48,6 +48,8 @@ export type UtxoChainId =
   | KnownChainIds.DogecoinMainnet
   | KnownChainIds.LitecoinMainnet
 
+export type SvmChainId = KnownChainIds.SolanaMainnet
```
preferably-blocking: remove me - this makes sense for EVM ChainIds as an intersection of EVM ChainIds, however, for Solana, SolanaMainnet is the only supported so we should either use `KnownChainIds.SolanaMainnet` directly (preferred for the sake of simplicity), or, if we really want to early abstract this for different SVM ChainIds, than this should be `SolanaChainId` as an alias for the time being

gomesalexandre commented on .env.base:
```diff
@@ -61,13 +61,14 @@ REACT_APP_FEATURE_DYNAMIC_LP_ASSETS=false
 
 # swapper feature flags - other .env files will override these
-REACT_APP_FEATURE_CHAINFLIP=false
+REACT_APP_FEATURE_CHAINFLIP_SWAP=false
```
Update `REACT_APP_FEATURE_CHAINFLIP_DCA` to `REACT_APP_FEATURE_CHAINFLIP_SWAP_DCA` across all three envs too

NeOMakinG replied:
Changed in [4e76e81](https://github.com/shapeshift/web/pull/8120/commits/4e76e8115223d91bd602f7d9ceed80c94c77ea15)

gomesalexandre commented on src/state/helpers.ts:
```diff
@@ -38,9 +48,12 @@ export const getEnabledSwappers = (
+      ChainflipSwap &&
+      (!isCrossAccountTrade || isCrossAccountTradeSupported(SwapperName.Chainflip)),
+    [SwapperName.Jupiter]:
+      JupiterSwap && (!isCrossAccountTrade || isCrossAccountTradeSupported(SwapperName.Jupiter)),
```
i.e this should be something like
```suggestion
      JupiterSwap && (!isCrossAccountTrade || (!isSolBuyAssetId && isCrossAccountTradeSupported(SwapperName.Jupiter))),
```

Where `isSolBuyAssetId` is passed to `getEnabledSwappers` - this way, we can remove all the additional `isManualReceiveAddress` plumbing, and most importantly ensure that we stick to current swapper flow i.e no wallet connected is able to yield rates regardless of cross-account support, while connecting a wallet yields only rates for supported cross-account routes

NeOMakinG replied:
Fixed in [0001e53](https://github.com/shapeshift/web/pull/8120/commits/0001e5316ccd4395724ab75e36a101390df9b1d8)

gomesalexandre replied:
Much love very cleaner ser!

gomesalexandre reviewed (commented):
Conceptual review, untested at runtime just yet but solid PR ser.

A bunch of cleanup comments and q.s, will test at runtime when all tackled 🙏🏽

NeOMakinG commented on packages/chain-adapters/src/solana/SolanaChainAdapter.ts:
```diff
@@ -62,6 +66,8 @@ import { toAddressNList, toRootDerivationPath } from '../utils'
 import { assertAddressNotSanctioned } from '../utils/validateAddress'
 import { microLamportsToLamports } from './utils'
 
+export const svmChainIds = [KnownChainIds.SolanaMainnet] as const
```
We probably don't need it but I think we should keep it, it doesn't hurt and it's ISO with some other chains family

gomesalexandre replied:
As dicussed, happy to keep this guy for conschistency 🙏🏽

NeOMakinG commented on packages/chain-adapters/src/solana/SolanaChainAdapter.ts:
```diff
@@ -449,8 +460,13 @@ export class ChainAdapter implements IChainAdapter<KnownChainIds.SolanaMainnet>
     const { to, chainSpecific } = input
     const { from, tokenId, instructions = [] } = chainSpecific
 
-    if (!to) throw new Error('to is required')
```
Added it back in [4e76e81](https://github.com/shapeshift/web/pull/8120/commits/4e76e8115223d91bd602f7d9ceed80c94c77ea15)  with an `isUndefined` check, in both cases we provide the `to`, but `!to` is evaluating to false if it's an empty string

NeOMakinG commented on packages/chain-adapters/src/solana/SolanaChainAdapter.ts:
```diff
@@ -449,8 +460,13 @@ export class ChainAdapter implements IChainAdapter<KnownChainIds.SolanaMainnet>
 
-    if (!to) throw new Error('to is required')
-    if (!input.value) throw new Error('value is required')
+    const estimationInstructions = [...instructions]
```
If we don't spread, we mutate the `instructions` ref, resulting in adding 2 times the instructions, in the fee estimations and in the final build, making the TX reverting

NeOMakinG commented on packages/chain-adapters/src/solana/SolanaChainAdapter.ts:
```diff
@@ -521,18 +546,46 @@ export class ChainAdapter implements IChainAdapter<KnownChainIds.SolanaMainnet>
+    }
+  }
+
+  private async buildTokenTransferInstructions({
```
Done in [4e76e81](https://github.com/shapeshift/web/pull/8120/commits/4e76e8115223d91bd602f7d9ceed80c94c77ea15)

NeOMakinG commented on packages/chain-adapters/src/solana/SolanaChainAdapter.ts:
```diff
@@ -557,6 +610,12 @@ export class ChainAdapter implements IChainAdapter<KnownChainIds.SolanaMainnet>
     }
   }
 
+  public async getTxStatus(tx: unchained.solana.Tx, pubkey: string): Promise<unchained.TxStatus> {
+    const parsedTx = await this.parseTx(tx, pubkey)
+
+    return parsedTx.status
+  }
+
```
It's used in `checkSolanaSwapStatus`, we need it along with the parser to get the status, I would keep it here

NeOMakinG commented on packages/chain-adapters/src/solana/SolanaChainAdapter.ts:
```diff
@@ -572,4 +631,54 @@ export class ChainAdapter implements IChainAdapter<KnownChainIds.SolanaMainnet>
+    return await this.connection.getMultipleAccountsInfo(addresses.map(key => new PublicKey(key)))
+  }
+
+  public async getSolanaAddressLookupTableAccountsInfo(
```
Makes sense, changed in [4e76e81](https://github.com/shapeshift/web/pull/8120/commits/4e76e8115223d91bd602f7d9ceed80c94c77ea15)

NeOMakinG commented on packages/chain-adapters/src/solana/SolanaChainAdapter.ts:
```diff
@@ -572,4 +631,54 @@ export class ChainAdapter implements IChainAdapter<KnownChainIds.SolanaMainnet>
+    }, new Array<AddressLookupTableAccount>())
+  }
+
+  public async getAddressLookupTableAccounts(
```
Makes sense, changed in [4e76e81](https://github.com/shapeshift/web/pull/8120/commits/4e76e8115223d91bd602f7d9ceed80c94c77ea15)

NeOMakinG commented on packages/swapper/src/swappers/ChainflipSwapper/swapperApi/getTradeQuote.ts:
```diff
@@ -253,7 +253,7 @@ export const _getTradeQuote = async (
     const feeData = await getFeeData()
 
-    if (isStreaming && !deps.config.REACT_APP_FEATURE_CHAINFLIP_DCA) {
+    if (isStreaming && !deps.config.REACT_APP_FEATURE_CHAINFLIP_SWAP_DCA) {
```
Changed in [4e76e81](https://github.com/shapeshift/web/pull/8120/commits/4e76e8115223d91bd602f7d9ceed80c94c77ea15)

NeOMakinG commented on packages/swapper/src/swappers/JupiterSwapper/JupiterSwapper.ts:
```diff
@@ -0,0 +1,33 @@
+          return jupiterSupportedChainIds.includes(chainId)
+        })
+        .map(asset => asset.assetId),
+    )
```
Changed in [4e76e81](https://github.com/shapeshift/web/pull/8120/commits/4e76e8115223d91bd602f7d9ceed80c94c77ea15)

NeOMakinG commented on packages/swapper/src/swappers/JupiterSwapper/swapperApi/getTradeQuote.ts:
```diff
@@ -0,0 +1,231 @@
+
+  const jupiterUrl = deps.config.REACT_APP_JUPITER_API_URL
+
+  if (isManualReceiveAddress && buyAsset.assetId === solAssetId) {
```
it's already blocked as rate time as getTradeQuote === getTradeRate for Jupiter

I would not add any specific cases in getEnabledSwappers and isCrossAccountTradeSupported as it should belong to the swapper itself, those checks are pretty dumb, for the sake of simplicity and plumbing, I would keep specific cases into each swapper handlers instead of getting them outside of their scopes

gomesalexandre replied:
>  for the sake of simplicity and plumbing

Main concern comes re: plumbing this all the way through, vs. `getEnabledSwappers` taking an additional param in which would be a one-liner in two places, not confident with the amount of boilerplate this brings and possible risks vs. passing an additional prop in one function call only

gomesalexandre replied:
Also re: `getTradeQuote === getTradeRate`, the functions may or may not be the same, but the quotes are two different ones (rate quote at input time vs. quote quote at confirm time) - FYI once you do connect a wallet, cross-account checks will kick in and a new rate will be gotten 🙏🏽 meaning you can get into a state where you get all swappers for a given pair, but once you connect and select some cross-accout pair, the previous selection should be void.

This is part of the swapper flow as it is, and I believe we should keep it as simple as-is, the only variable here as opposed to good ol' lookup in constants for "does this swapper support cross-account trade" is the addition of the notion of `isSolanaBuyAssetId`

NeOMakinG replied:
Leverage `getEnabledSwappers` in [0001e53](https://github.com/shapeshift/web/pull/8120/commits/0001e5316ccd4395724ab75e36a101390df9b1d8)

NeOMakinG commented on packages/swapper/src/swappers/JupiterSwapper/swapperApi/getTradeQuote.ts:
```diff
@@ -0,0 +1,231 @@
+        const sellAdapter = deps.assertGetSolanaChainAdapter(sellAsset.chainId)
+        const getFeeDataInput: GetFeeDataInput<KnownChainIds.SolanaMainnet> = {
+          // Simulates a self-send, since we don't know the to just yet at this stage
+          to: input.sendAddress!,
```
Unfortunately we cannot calculate the compute units before getting the quote, the only fees estimation we can do is a regular send, which will change execution time when we have all the jupiter instructions including compute units

Updated to use the `to` from the input instead but it doesn't change so much tbh as it's a regular send

We might have a small fee difference at this point because of the compute units that we can't calculate there...

gomesalexandre replied:
> Unfortunately we cannot calculate the compute units before getting the quote
> Also at rate time receive address is not defined as a part of the split PR we merged

This is `getTradeQuote` so we should definitely be able to? I think the main issue here may be WRT reusing the exact same logic for trades and quotes for Jupiter, looks like quotes should be able to leverage the full Tx data including full instructions set?

NeOMakinG replied:
It might be doable, but I think we shouldn't call the quote endpoint at rate time so we have to split them

Let me try something tomorrow:
- Adding the `getJupiterSwapInstructions` execution at quote and rate time so we can get a more realistic fees estimation

gomesalexandre replied:
Sounds much saner
> we shouldn't call the quote endpoint at rate time so we have to split them

That's absolutely the goal 🙏🏽 basically, if the logic really is the same for both, it's fine to abuse a quote endpoint to get a rate (a quote is really just a rate with defined BIP44 params).
But if they happen to differ even slightly, their logic should be consolidated somewhere sane and reused, so that either rate or quote can have some additional/different functionality (typically things like getting *real* fees estimation at quote time with real calldata is a good contender for this different logic, since depending on the swapper we may not know the `to` address, or approvals may not be granted which may result in reverts etc)

gomesalexandre replied:
Closing as split has been tackled proper

NeOMakinG commented on packages/swapper/src/swappers/JupiterSwapper/endpoints.ts:
```diff
@@ -0,0 +1,144 @@
+          const sellAdapter = assertGetSolanaChainAdapter(step.sellAsset.chainId)
+          const getFeeDataInput: GetFeeDataInput<KnownChainIds.SolanaMainnet> = {
+            // Simulates a self-send
+            to: from,
```
Changed to `''` as we are calculating the exact fees at this point and we have the instructions, nice catch

[4e76e81](https://github.com/shapeshift/web/pull/8120/commits/4e76e8115223d91bd602f7d9ceed80c94c77ea15)

gomesalexandre commented on packages/chain-adapters/src/solana/SolanaChainAdapter.ts:
```diff
@@ -466,6 +467,7 @@ export class ChainAdapter implements IChainAdapter<KnownChainIds.SolanaMainnet>
       chainSpecific.addressLookupTableAccounts ?? [],
     )
 
+    if (isUndefined(input.to)) throw new Error(`${this.getName()}ChainAdapter: value is required`)
```
```suggestion
    if (isUndefined(input.to)) throw new Error(`${this.getName()}ChainAdapter: to is required`)
```

NeOMakinG replied:
Fixed in [0001e53](https://github.com/shapeshift/web/pull/8120/commits/0001e5316ccd4395724ab75e36a101390df9b1d8)

gomesalexandre commented on src/state/slices/tradeQuoteSlice/selectors.ts:
```diff
@@ -70,7 +70,7 @@ const selectEnabledSwappersIgnoringCrossAccountTrade = createSelector(
   selectFeatureFlags,
   featureFlags => {
     // cross account trade logic is irrelevant here, so we can set the flag to false here
```
flags vs. flag previously

NeOMakinG replied:
[9c66e61](https://github.com/shapeshift/web/pull/8120/commits/9c66e614642dfa0f9d972d0b6d333294c3973807)

gomesalexandre commented on packages/swapper/src/swappers/JupiterSwapper/endpoints.ts:
```diff
@@ -0,0 +1,144 @@
+
+    if (isManualReceiveAddress && step.buyAsset.assetId === solAssetId) {
+      throw Error('Manual receive address is not supported for SOL')
+    }
```
@NeOMakinG with the `isCrossAccountTrade` logic removed, do we want to remove this guy? This is now handled by web the conventional way

https://github.com/shapeshift/web/blob/0001e5316ccd4395724ab75e36a101390df9b1d8/packages/swapper/src/swappers/JupiterSwapper/endpoints.ts#L41-L45

gomesalexandre reviewed (commented):
Looking much better now ser - barring improvements re: trades and quotes and cross-account checks seemingly wrong (using SOL sell asset instead of buy asset), diff is looking gucci.
Runtime pass up next!

NeOMakinG commented on src/state/apis/swapper/swapperApi.ts:
```diff
@@ -56,11 +56,16 @@ export const swapperApi = createApi({
           quoteOrRate,
         } = tradeQuoteInput
 
+        const isSolSellAssetId = sellAsset.assetId === solAssetId
```
... fml 

[9c66e61](https://github.com/shapeshift/web/pull/8120/commits/9c66e614642dfa0f9d972d0b6d333294c3973807)

gomesalexandre commented on packages/swapper/src/swappers/JupiterSwapper/swapperApi/getTradeQuote.ts:
```diff
@@ -0,0 +1,439 @@
+      case CHAIN_NAMESPACE.Solana: {
+        const sellAdapter = deps.assertGetSolanaChainAdapter(sellAsset.chainId)
+        const getFeeDataInput: GetFeeDataInput<KnownChainIds.SolanaMainnet> = {
+          to: receiveAddress ?? sendAddress,
```
This doesn't look correct as `receiveAddress` is always undefined according to type, and `sendAddress` is `never`

gomesalexandre replied:
Looking into quote method, this also just happens to be a PDA send, looks like what we're after is `to: ''` here?

gomesalexandre commented on packages/swapper/src/swappers/JupiterSwapper/swapperApi/getTradeQuote.ts:
```diff
@@ -0,0 +1,439 @@
+          to: receiveAddress ?? sendAddress,
+          value: sellAmount,
+          chainSpecific: {
+            from: input.sendAddress!,
```
This doesn't look correct either as `sendAddress` is always undefined so non-null assertion won't do anything here either
<img width="472" alt="image" src="https://github.com/user-attachments/assets/540b5c22-7a74-484c-a4a0-7fb6a950cf19">

gomesalexandre replied:
Also tackled

gomesalexandre commented on packages/swapper/src/swappers/JupiterSwapper/swapperApi/getTradeQuote.ts:
```diff
@@ -0,0 +1,439 @@
+    )
+  }
+
+  if (!sendAddress) {
```
This doesn't look correct - a rate without a wallet connected is guaranteed to never have a `sendAddress`

gomesalexandre replied:
Also removed

gomesalexandre commented on packages/swapper/src/swappers/JupiterSwapper/swapperApi/getTradeQuote.ts:
```diff
@@ -0,0 +1,439 @@
+
+const tradeQuoteMetadata: Map<string, QuoteResponse> = new Map()
+
+export const getTradeRate = async (
```
preferably-blocking: extract me into a `getTradeRate.ts` method similar to other swappers

gomesalexandre replied:
Unapplicable anymore, closing

gomesalexandre commented on packages/swapper/src/swappers/JupiterSwapper/swapperApi/getTradeQuote.ts:
```diff
@@ -0,0 +1,439 @@
+  const getFeeData = async () => {
+    const { chainNamespace } = fromAssetId(sellAsset.assetId)
+
+    switch (chainNamespace) {
```
nitpick: switch not required here and in quote method

gomesalexandre replied:
Also tackled, tyvm!

gomesalexandre reviewed (changes requested):
Tested locally:

- Unable to get rates without wallet connected (see inline comments for the reason why) 🚫 

<img width="1228" alt="image" src="https://github.com/user-attachments/assets/367b18b0-e091-4ce7-bc6e-1bf763169871">

### Wallet connected - no custom receive addy

- Able to get SOL sends rates with wallet connected ✅ 

<img width="1374" alt="image" src="https://github.com/user-attachments/assets/3d83653b-f179-4d83-817a-201bc0cf646a">

- Able to get and execute SOL sell quote with wallet connected 🚫 

https://jam.dev/c/bb637228-1a2d-4b72-af70-0d8644693849

Tx simulation is failing on Solana popup

<img width="305" alt="image" src="https://github.com/user-attachments/assets/c31e8bf9-57ee-4686-9689-abcab2e57f92">

And actual Tx is failing too

<img width="497" alt="image" src="https://github.com/user-attachments/assets/9b5fb54a-0904-4321-ad3c-4e8531b2fbe1">

- Able to get SOL buy rates with wallet connected ✅ 

<img width="1405" alt="image" src="https://github.com/user-attachments/assets/d027a51c-2be9-43be-bf9e-69fa8816a745">

- Able to execute SOL buy quotes with wallet connected 🚫 

Same failure as above

https://jam.dev/c/0bed7f93-1d0e-4ab6-b8e3-977ee8c81457

### Custom receive buy addy

- Able to get SOL sends rates with wallet connected ✅ 

<img width="1507" alt="image" src="https://github.com/user-attachments/assets/eb6b9d0a-7b30-445a-b763-76be955ddbf5">

- Able to get and execute SOL sell quote with wallet connected 🚫 

Managed to get unchained to return a Txid once (though never seemed to hit the mempool), and got the same error as above afterwards

https://jam.dev/c/b69e29f8-ea76-4635-ba8f-7049c2040ba7

- Should be unable to get SOL buy rates with wallet connected ✅ (cross-account execution not available for SOL buys)

<img width="1044" alt="image" src="https://github.com/user-attachments/assets/c412cd7d-b0e3-4874-abb0-216c38506763">
<img width="995" alt="image" src="https://github.com/user-attachments/assets/12e61476-5400-4a37-a6a8-afd1f0f9ed39">

- Able to get and execute SOL buy quote with wallet connected 🚫 

https://jam.dev/c/1c567ab2-28ff-4cb4-b00f-dbe5a0478b21

### Misc

- Able to get rates for Wrapped SOL ✅ 

<img width="978" alt="image" src="https://github.com/user-attachments/assets/e3affced-4df6-4f81-822b-262f08abadc3">
<img width="1052" alt="image" src="https://github.com/user-attachments/assets/a7636a96-d665-429d-9eb3-abf9f7c43743">

<img width="623" alt="image" src="https://github.com/user-attachments/assets/1c830d30-48af-489b-83dd-d56950a872cb">
<img width="958" alt="image" src="https://github.com/user-attachments/assets/ae6d10be-f387-41c5-bdc1-db9b38e4fae4">

NeOMakinG commented:
@gomesalexandre we should be good for another review/runtime test 🙏

woodenfurniture commented on packages/swapper/src/swappers/JupiterSwapper/utils/helpers.ts:
```diff
@@ -0,0 +1,73 @@
+      `&platformFeeBps=${commissionBps}`,
+  )
+
+// @TODO: Add DAO's fee account
```
👀

woodenfurniture commented on packages/swapper/src/swappers/JupiterSwapper/utils/constants.ts:
```diff
@@ -0,0 +1,16 @@
+export const SOLANA_RANDOM_ADDRESS = '2zHKF6tqam3tnNFPK2E9nBDkV7GMXnvdJautmzqQdn8A'
+
+// Jupiter use 40% as a compute unit margin while calculating them, some TX reverts without this
+export const JUPITER_COMPUTE_UNIT_MARGIN_PERCENT = 1.4
```
[nitpick]: This looks like a scalar not a percentage

<img src="https://github.com/user-attachments/assets/9c010436-330b-4d59-a79f-70a84c924973" width="200">

woodenfurniture reviewed (changes requested):
# Testing notes

Looks like some flakienss that needs to be addressed before I go deep on testing.

### Unable to get SOL -> Wrapped SOL rate ❌
<img src="https://github.com/user-attachments/assets/b39c4b0e-542a-474f-a56f-cff5a0d557dc" width="200">

### Unable to execute SOL -> USDC trade ❌
https://jam.dev/c/0ab18444-b0f6-4702-b481-2691a88e3a5f
<img src="https://github.com/user-attachments/assets/360ddecd-09ec-4e68-849b-d737fed2f3f2" width="200">

Tried 4x more times, saw 3 failures as above and eventually got a trade to go through
<img src="https://github.com/user-attachments/assets/3525d543-24ed-4f53-9276-2ea3c09ef4d0" width="200">

<img src="https://github.com/user-attachments/assets/2f87a226-b7a2-4ec3-88fb-f98076ca17f7" width="200">

### Unable to execute USDC -> Wrapped SOL trade ❌
4 attempts:
https://jam.dev/c/2aa2cd51-dddf-4a86-b1d7-87ecb66e20f4

### Successful USDC -> SOL trade ✅
https://jam.dev/c/771cc054-79de-4631-8a61-bee44ef61737

gomesalexandre commented on packages/swapper/src/constants.ts:
```diff
@@ -105,7 +105,7 @@ const DEFAULT_LIFI_SLIPPAGE_DECIMAL_PERCENTAGE = '0.005' // .5%
 const DEFAULT_ARBITRUM_BRIDGE_SLIPPAGE_DECIMAL_PERCENTAGE = '0' // no slippage for Arbitrum Bridge, so no slippage tolerance
 const DEFAULT_CHAINFLIP_SLIPPAGE_DECIMAL_PERCENTAGE = '0.02' // 2%
-const DEFAULT_JUPITER_SLIPPAGE_DECIMAL_PERCENTAGE = '0.005' // .5%
+const DEFAULT_JUPITER_SLIPPAGE_DECIMAL_PERCENTAGE = '0.01' // .5%
```
Update comment too

gomesalexandre commented on packages/swapper/src/swappers/JupiterSwapper/endpoints.ts:
```diff
@@ -1,15 +1,433 @@
+      )
+    }
+
+    const maybePriceResponse = await getJupiterPrice({
```
💜

gomesalexandre commented on packages/swapper/src/swappers/JupiterSwapper/swapperApi/getTradeQuote.ts:
```diff
@@ -0,0 +1,439 @@
+import type { QuoteResponse } from '../models/QuoteResponse'
+import { getJupiterQuote, getJupiterSwapInstructions, isSupportedChainId } from '../utils/helpers'
+
+const tradeQuoteMetadata: Map<string, QuoteResponse> = new Map()
```
Tackled in latest diff, resolving

gomesalexandre commented on packages/swapper/src/swappers/JupiterSwapper/swapperApi/getTradeQuote.ts:
```diff
@@ -0,0 +1,439 @@
+    )
+  }
+
+  const maybeQuoteResponse = await getJupiterQuote({
```
Tyvm for tackling this ser

gomesalexandre commented on packages/swapper/src/swappers/JupiterSwapper/endpoints.ts:
```diff
@@ -1,15 +1,433 @@
+    const getFeeData = async () => {
+      const sellAdapter = deps.assertGetSolanaChainAdapter(sellAsset.chainId)
+      const getFeeDataInput: GetFeeDataInput<KnownChainIds.SolanaMainnet> = {
+        // used as a placeholder for the sake of loosely estimating fees
+        to: SOLANA_RANDOM_ADDRESS,
```
💜

gomesalexandre reviewed (changes requested):
Can confirm the flakiness spotted by @woodenfurniture here: https://gist.github.com/gomesalexandre/bbf9f37e179b1ecc31e7505171ed08bb

Main takeaways of this testing:

- SOL <-> Wrapped SOL rates are borked (and their execution would probably be too if we could get to this stage with current diff)
- Execution in general is flaky - either with reverts during simulation, or with no reverts, but not hitting the mempool seemingly 
- Our ws seems flaky, though not related to this PR 

q1. WRT execution @NeOMakinG: are we sure that the current way of keeping the rate from rate and only adding additional props to it yields consistent successes?
To me, it seems like getting a final quote the usual way with the right compute units is the way, and getting said additional data at quote time *while* keeping the original rate data is a recipe for flakiness, but this may or may not be the correct way for Solana, not sure.
q2.: how much of a lift would it be to add auto-slippige here for Jupiter as a first swapper to ensure better reliability (and rates while at that), and do we want to do that in this PR to mitigate failures here?

NeOMakinG commented:
- I disabled explicitly swapping between wrappedSOL to any other asset and any asset to wrappedSOL as it's considered as `SOL` from the Jupiter point of view, I updated the PR description because we may need to provide an unwrap feature to our users as Jupiter does
- Fixed all the last comments

⚠️ I can't reproduce the USDC => SOL and SOL => USDC bugs, I tried a couple of swaps and none of them was not picked by the mempool, all the swaps were successful with good rates

q1 @gomes : Also tried to wait around 40 minutes after fetching the quote, it did fail, so I decided to fetch the prices again in the getTradeQuote handler in [5780800] (https://github.com/shapeshift/web/pull/8120/commits/578080067d966c0b382860fda2c36afef6621d92)

Can you please guys test again and tell me if you see any more flaky behaviors? 

q2: From the Jupiter point of view, the list is not so big: Update <SharedSlippagePopover>` to be accepting `Auto, Default, Custom`, add a prod `acceptAuto`, that should be enabled for the jupiter swapper
Ensure all mathematics done with slippage are handling the "auto" mode, specially for the jupiter swaper

But please, we should try this in a follow up if we really want to, not on this particular PR which starts to be big

NeOMakinG reviewed (commented):
Answered all the comments

gomesalexandre commented on packages/swapper/src/swappers/JupiterSwapper/swapperApi/getTradeQuote.ts:
```diff
@@ -0,0 +1,273 @@
+    fromAddress: sendAddress,
+    toAddress: isCrossAccountTrade ? destinationTokenAccount?.toString() : undefined,
+    rawQuote: quoteResponse,
+    wrapAndUnwrapSol: buyAsset.assetId === wrappedSolAssetId ? false : true,
```
@NeOMakinG shouldn't this go away as discussed?

gomesalexandre commented on packages/swapper/src/swappers/JupiterSwapper/swapperApi/getTradeQuote.ts:
```diff
@@ -0,0 +1,273 @@
+    fromAddress: sendAddress,
+    toAddress: isCrossAccountTrade ? destinationTokenAccount?.toString() : undefined,
+    rawQuote: quoteResponse,
+    wrapAndUnwrapSol: buyAsset.assetId === wrappedSolAssetId ? false : true,
```
This can also go away now that wrapped SOL is unsupported 
```suggestion
    wrapAndUnwrapSol: true,
```

NeOMakinG replied:
Done in [561cd08](https://github.com/shapeshift/web/pull/8120/commits/561cd0854969e274843124d716a24964dec5892e)

gomesalexandre reviewed (approved):
Stamping as I ended up with the simulation failed on Phantom / Tx not broadcasted to the mempool only once, and couldn't repro after (which is why there is none in this Jam) despite many attempts). E2E testing below:

https://jam.dev/c/4bae293f-4fb2-47eb-85ab-1b53d231642e

Still noticing some Txs rarely failing for a diff. reason, though increasing slippeedge seems to solve it, so not too concerned about this as we will fast-follow with auto-slippage (https://github.com/shapeshift/web/issues/8193)

https://explorer.solana.com/tx/ajy2v17CFGZiuXvwC8NvKZqDeeeqgYLtGww39aRXBREwH4Bo9o2ZDGz859Fwy4SmKwu5GrUFHhjd59p6UD9XAHq

@NeOMakinG if you feel like it as a fast-follow, the ETA as "<x> ms" where x < 100 is still looking very odd to me, looks like it got lost in translation in previous review (https://gist.github.com/gomesalexandre/bbf9f37e179b1ecc31e7505171ed08bb) but would be nice to tackle as a one-liner fast-follow, or in this PR if you feel like it, to make it no ETA similar to most EVM swappers (see 0x and Portals in the screenshot below missing it altogether, since swap is done within a block i.e effectively instantly, vs. Jup showing some ETA in the realm of ms):

<img width="493" alt="image" src="https://github.com/user-attachments/assets/1deeb863-63c6-4f91-8a6b-6de849009dc2">
<img width="434" alt="image" src="https://github.com/user-attachments/assets/c2026a07-84ab-442b-8530-03ccfd58a4c5">

NeOMakinG commented:
> Stamping as I ended up with the simulation failed on Phantom / Tx not broadcasted to the mempool only once, and couldn't repro after (which is why there is none in this Jam) despite many attempts). E2E testing below:
> 
> https://jam.dev/c/4bae293f-4fb2-47eb-85ab-1b53d231642e
> 
> Still noticing some Txs rarely failing for a diff. reason, though increasing slippeedge seems to solve it, so not too concerned about this as we will fast-follow with auto-slippage (#8193)
> 
> https://explorer.solana.com/tx/ajy2v17CFGZiuXvwC8NvKZqDeeeqgYLtGww39aRXBREwH4Bo9o2ZDGz859Fwy4SmKwu5GrUFHhjd59p6UD9XAHq
> 
> @NeOMakinG if you feel like it as a fast-follow, the ETA as " ms" where x < 100 is still looking very odd to me, looks like it got lost in translation in previous review (https://gist.github.com/gomesalexandre/bbf9f37e179b1ecc31e7505171ed08bb) but would be nice to tackle as a one-liner fast-follow, or in this PR if you feel like it, to make it no ETA similar to most EVM swappers (see 0x and Portals in the screenshot below missing it altogether, since swap is done within a block i.e effectively instantly, vs. Jup showing some ETA in the realm of ms):
> 
> <img alt="image" width="493" src="https://private-user-images.githubusercontent.com/17035424/390623591-1deeb863-63c6-4f91-8a6b-6de849009dc2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI3ODEzNDYsIm5iZiI6MTczMjc4MTA0NiwicGF0aCI6Ii8xNzAzNTQyNC8zOTA2MjM1OTEtMWRlZWI4NjMtNjNjNi00ZjkxLThhNmItNmRlODQ5MDA5ZGMyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTI4VDA4MDQwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIwMTI5MGI3MGY3MWI0NmU2OTM5MmU5NGYyNjlkYzAyMGRhYWIzNDM2MTc1ZTI4MjA2ZjEzMWFiOTYxZjQzNTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.rvX4qDgTmV9SsA0KQzlQLlnxH33aOk-tt1-tuwWJ_Ew"> <img alt="image" width="434" src="https://private-user-images.githubusercontent.com/17035424/390623662-c2026a07-84ab-442b-8530-03ccfd58a4c5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI3ODEzNDYsIm5iZiI6MTczMjc4MTA0NiwicGF0aCI6Ii8xNzAzNTQyNC8zOTA2MjM2NjItYzIwMjZhMDctODRhYi00NDJiLTg1MzAtMDNjY2ZkNThhNGM1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTI4VDA4MDQwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY3OWE1MDM4NTBkN2ViNjg5MDhhZTAxM2UwNWExYTNkN2ViNDNmOTM1ZDIwMzU3OTczY2EyNmM4OTE1YjU5M2EmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.EBrBCa31OnD4GYmHim_uxGLJSM_cl_HlU6TBTDd1DUE">

Removed the time for jupiter in [561cd08](https://github.com/shapeshift/web/pull/8120/commits/561cd0854969e274843124d716a24964dec5892e)

woodenfurniture reviewed (approved):
Confirmed swaps are looking much more stable now.

https://jam.dev/c/0f249b0c-7c8d-47b1-9a67-060da1fb2e16
https://jam.dev/c/6e14d3dd-137f-4beb-8cec-b7994d786d5e

Lets get this in and enjoy the follow-ups when they come 🎉
